File: mnemefusion-core/examples/basic_usage.rs
//! Basic usage example for MnemeFusion
//!
//! This example demonstrates the core functionality:
//! - Creating/opening a database
//! - Adding memories with embeddings
//! - Retrieving memories by ID
//! - Searching by semantic similarity
//! - Querying memories by time (recent, range)
//! - Adding causal relationships between memories
//! - Querying causal chains (causes and effects)
//! - Using metadata
//! - Deleting memories
//! - Persistence across restarts
//!
//! Run with: cargo run --example basic_usage

use mnemefusion_core::{Config, MemoryEngine};
use std::collections::HashMap;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("MnemeFusion Basic Usage Example\n");

    // Open or create a database
    println!("Opening database at ./example_brain.mfdb");
    let engine = MemoryEngine::open("./example_brain.mfdb", Config::default())?;

    // Show initial count
    let initial_count = engine.count()?;
    println!("Initial memory count: {}\n", initial_count);

    // Add some memories with simple embeddings
    println!("Adding memories...");

    let memories = vec![
        "Project deadline moved to March 15th",
        "Team meeting scheduled for tomorrow at 2pm",
        "Budget approved for Q2 expansion",
        "New hire starts next Monday",
        "Code review completed successfully",
    ];

    let mut memory_ids = Vec::new();

    for (idx, content) in memories.iter().enumerate() {
        // In a real application, you'd use a proper embedding model
        // For this example, we'll use simple dummy embeddings
        let embedding = vec![(idx as f32) * 0.1; 384];

        // Add metadata
        let mut metadata = HashMap::new();
        metadata.insert("example".to_string(), "basic_usage".to_string());
        metadata.insert("index".to_string(), idx.to_string());

        let id = engine.add(
            content.to_string(),
            embedding,
            Some(metadata),
            None,
            None,
            None,
        )?;
        println!("  Added: {} (ID: {}...)", content, &id.to_string()[..8]);
        memory_ids.push(id);
    }

    // Show updated count
    let count = engine.count()?;
    println!("\nTotal memories: {}", count);

    // Retrieve a specific memory
    println!("\nRetrieving first memory...");
    if let Some(memory) = engine.get(&memory_ids[0])? {
        println!("  Content: {}", memory.content);
        println!("  Embedding dimension: {}", memory.embedding.len());
        println!("  Metadata: {:?}", memory.metadata);
        println!(
            "  Created: {} seconds since epoch",
            memory.created_at.as_unix_secs()
        );
    }

    // List all memory IDs
    println!("\nAll memory IDs:");
    let all_ids = engine.list_ids()?;
    for id in &all_ids {
        println!("  {}", id);
    }

    // Search for similar memories
    println!("\nSearching for memories similar to 'meeting' topic...");
    // In a real application, you'd embed the query with the same model
    // For this example, we use a query vector similar to our second memory (team meeting)
    let query_embedding = vec![0.1; 384]; // Similar to idx=1 memory
    let results = engine.search(&query_embedding, 3, None, None)?;

    println!("  Found {} results:", results.len());
    for (idx, (memory, similarity)) in results.iter().enumerate() {
        println!(
            "    {}. [Similarity: {:.3}] {}",
            idx + 1,
            similarity,
            memory.content
        );
    }

    // Temporal queries
    println!("\nQuerying most recent memories...");
    let recent = engine.get_recent(3, None)?;
    println!("  3 most recent memories:");
    for (idx, (memory, timestamp)) in recent.iter().enumerate() {
        println!(
            "    {}. [{}] {}",
            idx + 1,
            timestamp.as_unix_secs() as u64,
            memory.content
        );
    }

    // Time range query
    use mnemefusion_core::Timestamp;
    let now = Timestamp::now();
    let one_hour_ago = now.subtract_days(0); // For demo, just use now as the range
    println!("\nQuerying memories from the last session...");
    let range_results = engine.get_range(one_hour_ago.subtract_days(1), now, 10, None)?;
    println!("  Found {} memories in time range", range_results.len());

    // Causal relationships
    println!("\nAdding causal relationships...");
    // memory_ids[0] = "Project deadline moved"
    // memory_ids[1] = "Team meeting scheduled"
    // memory_ids[2] = "Budget approved"

    // Create causal chain: deadline → meeting → budget
    engine.add_causal_link(
        &memory_ids[0],
        &memory_ids[1],
        0.85,
        "Deadline change triggered team meeting".to_string(),
    )?;

    engine.add_causal_link(
        &memory_ids[1],
        &memory_ids[2],
        0.75,
        "Meeting led to budget approval".to_string(),
    )?;

    println!("  Added 2 causal links");

    // Query causal effects
    println!("\nQuerying causal effects of deadline change...");
    let effects = engine.get_effects(&memory_ids[0], 2)?;
    println!("  Found {} causal paths:", effects.paths.len());
    for (idx, path) in effects.paths.iter().enumerate() {
        println!(
            "    {}. Path length: {}, confidence: {:.2}",
            idx + 1,
            path.memories.len(),
            path.confidence
        );
    }

    // Query causal causes
    println!("\nQuerying causes of budget approval...");
    let causes = engine.get_causes(&memory_ids[2], 2)?;
    println!("  Found {} causal paths backward:", causes.paths.len());
    for (idx, path) in causes.paths.iter().enumerate() {
        println!(
            "    {}. Path length: {}, confidence: {:.2}",
            idx + 1,
            path.memories.len(),
            path.confidence
        );
    }

    // Entity extraction and queries
    println!("\nEntity extraction (automatic from memory content)...");
    println!("  Entities were automatically extracted when memories were added");

    // List all entities
    let all_entities = engine.list_entities()?;
    println!("  Found {} unique entities:", all_entities.len());
    for entity in &all_entities {
        println!("    - {} ({} mentions)", entity.name, entity.mention_count);
    }

    // Query memories by entity
    if !all_entities.is_empty() {
        let example_entity = &all_entities[0];
        println!(
            "\nQuerying memories that mention '{}'...",
            example_entity.name
        );
        let entity_memories = engine.get_entity_memories(&example_entity.name)?;
        println!("  Found {} memories:", entity_memories.len());
        for (idx, memory) in entity_memories.iter().enumerate() {
            println!("    {}. {}", idx + 1, memory.content);
        }
    }

    // Query entities in a specific memory
    println!("\nQuerying entities in first memory...");
    let entities_in_memory = engine.get_memory_entities(&memory_ids[0])?;
    println!("  Found {} entities:", entities_in_memory.len());
    for entity in entities_in_memory {
        println!("    - {}", entity.name);
    }

    // Delete a memory
    println!("\nDeleting the third memory...");
    let deleted = engine.delete(&memory_ids[2], None)?;
    if deleted {
        println!("  Successfully deleted");
        println!("  Remaining memories: {}", engine.count()?);
    }

    // Close the database (saves all indexes)
    println!("\nClosing database...");
    engine.close()?;

    println!("\n✓ Example completed successfully!");
    println!("The database file 'example_brain.mfdb' has been created.");
    println!("Run this example again to see persistence in action.");

    Ok(())
}

File: mnemefusion-python/examples/basic_usage.py
"""
Basic usage example for MnemeFusion Python bindings

This example demonstrates:
- Creating a memory database
- Adding memories with embeddings
- Semantic search
- Intelligent query with intent classification
- Causal relationships
- Entity tracking
"""

import mnemefusion
import random


def random_embedding(dim=384):
    """Generate a random embedding for demonstration purposes"""
    return [random.random() for _ in range(dim)]


def main():
    print("=" * 60)
    print("MnemeFusion Python Example")
    print("=" * 60)
    print()

    # Create or open a memory database
    print("Opening memory database...")
    memory = mnemefusion.Memory("example_python.mfdb", config={
        "embedding_dim": 384,
        "entity_extraction_enabled": True,
    })

    print(f"Current memory count: {memory.count()}")
    print()

    # Add some memories
    print("Adding memories...")

    meeting_id = memory.add(
        "Team meeting scheduled for March 15th to discuss Project Alpha",
        random_embedding(),
        metadata={"type": "meeting", "project": "Alpha"}
    )
    print(f"  Added meeting: {meeting_id}")

    cancel_id = memory.add(
        "Meeting cancelled due to scheduling conflict with stakeholder review",
        random_embedding(),
        metadata={"type": "cancellation"}
    )
    print(f"  Added cancellation: {cancel_id}")

    notes_id = memory.add(
        "Alice provided detailed feedback on the API design for Project Alpha",
        random_embedding(),
        metadata={"type": "note", "author": "Alice"}
    )
    print(f"  Added notes: {notes_id}")

    print(f"\nTotal memories: {memory.count()}")
    print()

    # Add a causal relationship
    print("Adding causal relationship...")
    memory.add_causal_link(
        cancel_id,
        meeting_id,
        0.95,
        "Cancellation directly caused the meeting change"
    )
    print("  Linked: cancellation -> meeting")
    print()

    # Retrieve a memory
    print("Retrieving memory...")
    result = memory.get(meeting_id)
    if result:
        print(f"  Content: {result['content']}")
        print(f"  Created: {result['created_at']}")
        print(f"  Metadata: {result['metadata']}")
    print()

    # Semantic search
    print("Performing semantic search...")
    query_emb = random_embedding()
    search_results = memory.search(query_emb, top_k=3)

    print(f"  Found {len(search_results)} results:")
    for mem, score in search_results[:3]:
        print(f"    Score {score:.3f}: {mem['content'][:60]}...")
    print()

    # Intelligent query with intent classification
    print("Performing intelligent query...")
    print("  Query: 'Why was the meeting cancelled?'")

    intent, query_results = memory.query(
        "Why was the meeting cancelled?",
        random_embedding(),
        limit=5
    )

    print(f"  Intent: {intent['intent']} (confidence: {intent['confidence']:.2f})")
    print(f"  Results: {len(query_results)}")

    for mem, scores in query_results[:2]:
        print(f"    Fused score: {scores['fused_score']:.3f}")
        print(f"      Semantic: {scores['semantic_score']:.3f}")
        print(f"      Temporal: {scores['temporal_score']:.3f}")
        print(f"      Causal:   {scores['causal_score']:.3f}")
        print(f"      Entity:   {scores['entity_score']:.3f}")
        print(f"      Content: {mem['content'][:50]}...")
    print()

    # Get causal chains
    print("Exploring causal chains...")
    causes = memory.get_causes(meeting_id, max_hops=3)
    print(f"  Causes of meeting: {len(causes)} paths found")
    for i, path in enumerate(causes[:2]):
        print(f"    Path {i+1}: {len(path)} memories")
    print()

    # List entities
    print("Listing entities...")
    entities = memory.list_entities()
    print(f"  Found {len(entities)} entities:")
    for entity in entities[:5]:
        print(f"    - {entity['name']} (mentioned {entity['mention_count']} times)")
    print()

    # Delete a memory
    print("Deleting a memory...")
    deleted = memory.delete(notes_id)
    print(f"  Deleted: {deleted}")
    print(f"  New count: {memory.count()}")
    print()

    # Close the database
    print("Closing database...")
    memory.close()
    print("Done!")
    print()


def context_manager_example():
    """Example using context manager (with statement)"""
    print("=" * 60)
    print("Context Manager Example")
    print("=" * 60)
    print()

    with mnemefusion.Memory("example_context.mfdb") as memory:
        mem_id = memory.add(
            "This memory is added within a context manager",
            random_embedding()
        )
        print(f"Added memory: {mem_id}")
        print(f"Total: {memory.count()}")

    # Database is automatically closed when exiting the 'with' block
    print("Database automatically closed")
    print()


if __name__ == "__main__":
    main()
    context_manager_example()
